zng/docs/spec.md

```diff
@@ -753,3 +761,56 @@ For each ZNG primitive type, the following table describes:
 * [Protocol Buffers](https://developers.google.com/protocol-buffers)
 * [MessagePack](https://msgpack.org/index.html)
 * [gNMI](https://github.com/openconfig/reference/tree/master/rpc/gnmi)
+
+## Appendix B. Recommended Type Coercion rules
+
+> TBD: it might be better to put this in the ZQL docs
+
+While outside the scope of the ZNG format specification, we include here
+some suggested rules for converting types when mixed-type operations occur,
+e.g., adding an uint32 field to an int32 fields or aggregating a stream of
+mixed-type numeric values in a sum operator.
+
+The age old question is "does unsigned(-1) equal unsigned(maxint) and thus
+that signed -1 is larger than unsigned 1 in a coerced comparison?"
+The standard SQL specification goes so far as to avoid unsigned types altogether
+to avoid this confusion.  However, since unsigned types are prevalent
+in the real world, and we want ZNG to be a reliable and complete language-independent
+model for communicating structured data, ZNG embraces the unsigned type.
+
+Given the dynamic typing nature of ZNG streams (e.g., x in one record might
+be a uint8, in another an int64, and in still another, a string), type coercion
+is important for ergonomic use, and implementations are thus encouraged
+to handle mixed-type operations robustly.
+
+For systems that perform analytics directly on ZNG, the following coercion
+patterns are recommend for logical comparisons of numbers, arithmetic operations,
+or streaming aggregations over numbers:
+* For float32 and float64, the float32 is converted to float64.
+* For float32 and any integer type, the integer is converted to float32
+and any loss of precision causes no error.
+* For float64 and any integer type, the integer is converted to float64
+and any loss of precision causes no error.
+* For integers of same signed-ness but different widths, the smaller width
+type is converted to the wider type.
+* For any signed and unsigned integers smaller than 64 bits, the unsigned value
+is converted to the corresponding signed type if possible, and otherwise,
+both are converted to the widest signed type that will allow conversion of
+the unsigned value unless the unsigned value cannot be converted,
+in which case an overflow error occurs.   e.g., uint8(255) and int8(-1), are
+converted to int16(255) and int16(-1), but uint64(2^32) and any signed value
```

Comment:
```suggestion
converted to int16(255) and int16(-1), but uint64(1<<64-1) and any signed value
```

Reply:
This is now in in #1408 but the intent here was to talk aboubt uint64(maxint32+1) not maxuint64.